File: src/main/java/io/jjong/algorithm/linked/MergeList.java
package io.jjong.algorithm.linked;

/**
 * create on 2022/12/04. create by IntelliJ IDEA.
 *
 * <p> 클래스 설명 </p>
 * <p> {@link } and {@link }관련 클래스 </p>
 *
 * @author Jongsang Han(Henry)
 * @version 1.0
 * @see
 * @since 1.0
 */
public class MergeList {

  public static void main(String[] args){

    Node<Integer> node1 = new Node<>(1,
        new Node<>(2,
            new Node<>(5,
                new Node<>(9))));

    Node<Integer> node2 = new Node<>(3,
        new Node<>(4,
            new Node<>(7,
                new Node<>(8))));
    Node<Integer> mergeNode = mergeListNode(node1, node2);
    System.out.println(mergeNode);

  }


  public static Node<Integer> mergeListNode(Node<Integer> node1, Node<Integer> node2) {
    Node<Integer> dummyHead = new Node<>(0, null);
    Node<Integer> current = dummyHead;
    Node<Integer> p1 = node1, p2 = node2;

    while (p1 != null && p2 != null) {
      if (p1.data <= p2.data) {
        current.next = p1;
        p1 = p1.next;
      } else {
        current.next = p2;
        p2 = p2.next;
      }
      current = current.next;
    }
    current.next = p1 != null ? p1 : p2;
    return dummyHead.next;
  }
}

File: archive/src/main/kotlin/io/jjong/algorithm/linked/MergeList.kt
package io.jjong.algorithm.linked

/**
 * 정렬된 두 연결 리스트를 하나의 정렬된 리스트로 병합한다. Dummy Head 패턴 사용.
 */
object MergeList {
    fun mergeListNode(node1: Node<Int>?, node2: Node<Int>?): Node<Int>? {
        val dummyHead = Node<Int>(0, null)
        var current = dummyHead
        var p1 = node1
        var p2 = node2

        while (p1 != null && p2 != null) {
            if (p1.data!! <= p2.data!!) {
                current.next = p1
                p1 = p1.next
            } else {
                current.next = p2
                p2 = p2.next
            }
            current = current.next!!
        }
        current.next = p1 ?: p2
        return dummyHead.next
    }
}

fun main() {
    val node1 = Node(1, Node(2, Node(5, Node(9))))
    val node2 = Node(3, Node(4, Node(7, Node(8))))
    val mergeNode = MergeList.mergeListNode(node1, node2)
    println(mergeNode)
}
